### PR TITLE
Suggestions for what to edit and where for issue41

### PR DESCRIPTION
**Pull request by cculianu**
Issue #41 is slated to be implemented in this document.  I went ahead and implemented it already in the code for BCHN and this is what I suggest you specify here so as to be very clear and avoid subtle consensus errors.

Note that I added the requirement that to PARSE token data correctly, up to 32 MiB compactSize must be supported for COMMITMENT_LENGTH (which is the default limit for all compactSizes in Bitcoin).  Anything above this size fails to PARSE.  However commitments above 40 bytes will fail validation.  This point is important considering that we now have a distinction between:

- correctly parsed token data
- incorrectly parsed token data that just ends up as raw bytes into the locking script as its prefix

Why the 32MiB limit? This is the standard compactSize limit in Bitcoin used for all compactSizes (unless otherwise specified) and I feel it just makes the code do the "natural thing".  *For parsing*, we must accept such crazy token data -- even if it fails consensus later!  This is important because we really need to delineate what is a parse error and what is a consensus error.

### CODE REVIEW

**Comment by imaginaryusername** [comment]
Makes sense, but it sounds like a strange place for such a large number to be present - this will not be a DoS vector(since it's otherwise limited by the 10kb limit elsewhere) right?

**Comment by cculianu** [comment]
I understand your concern.  It won't be a DoS vector.  Other checks outside this ensure it can never be this large.  First of all a txn itself is limited to 1 MB anyway.  So in a real txn this never happen.  It will be limited to 1MB in practice. 

Note that the "outer enveloper" for the `scriptPubKey` already has a 32MiB check (it uses standard default `compactSize` encoding rules which limit it to 32MiB).  So this limit can never be hit anyway even because of that.. even if the 1MB txn size check never existed.

So.. why 32MiB?  You want to draw the line somewhere though for what is *never* going to be considered parseable and what may be in the future.  We can draw the line at 1MB, 10kb, 65kb, etc.  I suggested the existing "compactSize" limit which is 32MiB by default because it is what Bitcoin  does already in plenty of places (including in `scriptPubKey` which "wraps" this data already!!).

I feel this is a natural thing if you look at the code and the internals of how bitcoin operates.. but am open to other limits.

Again, note that this limit will never be hit in reality because other serialization and consensus rules around it ensure it can never happen.  It's just to delineate clear garbage from non-garbage.. and leave room for future expansion as well.

**Comment by imaginaryusername** [comment]
Got it. It might be useful to explain this in another couple lines in this PR, as the number may be scary to other onlookers who, like me, aren't nearly as knowledgeable about where the limits are.

**Comment by cculianu** [comment]
> Got it. It might be useful to explain this in another couple lines in this PR, as the number may be scary to other onlookers who, like me, aren't nearly as knowledgeable about where the limits are.

You have a good point...

File: readme.md [hunk 1]
```diff
@@ -620,7 +629,7 @@ Finally, this proposal's combined behavior preserves two codepoints in the Bitco
 
 ### Non-Fungible Token Commitment Length
 
-This specification limits the length of non-fungible token commitments to `40` bytes; this restrains excess growth of the UTXO set and reduces the resource requirements of typical wallets and indexers.
+This specification limits the length of non-fungible token commitments to `40` bytes for consensus (but to `33554432` bytes for serialization); this `40`-byte limit restrains excess growth of the UTXO set and reduces the resource requirements of typical wallets and indexers.
```
Comment (cculianu) [inline readme.md:632]:
```suggestion
This specification limits the length of non-fungible token commitments to `40` bytes for consensus (but to `33554432` bytes for serialization); this `40`-byte limit restrains excess growth of the UTXO set and reduces the resource requirements of typical wallets and indexers.

Note that the apparently-generous limit for serialization of `33554432` is not a DoS vector and is in-line with existing `compactSize` limits applied to bitcoin transaction components.  For instance, on the serialization layer, `scriptPubKey` already applies a `32 MiB` (`33554432` bytes) limit to `scriptPubKey` vector size. (even though the consensus limit is much smaller at `~1MB` for outputs, `10KB` when spending as an input).
```

**Comment by cculianu** [comment]
Ok, updated.

File: readme.md [hunk 2]
```diff
@@ -620,7 +629,9 @@ Finally, this proposal's combined behavior preserves two codepoints in the Bitco
...
+
+Note that the apparently-generous limit for serialization of `33554432` is not a DoS vector and is in-line with existing `compactSize` limits applied to bitcoin transaction components.  For instance, on the serialization layer, `scriptPubKey` already applies a `32 MiB` (`33554432` bytes) limit to `scriptPubKey` vector size. (even though the consensus limit is much smaller at `~1MB` for outputs, `10KB` when spending as an input).
```
Comment (cculianu) [inline readme.md:634]:
```suggestion
Note that the apparently-generous limit for serialization of `33554432` is not a DoS vector and is in-line with existing `compactSize` limits applied to bitcoin transaction components.  For instance, on the serialization layer, `scriptPubKey` already applies a `32 MiB` (`33554432` bytes) limit to `scriptPubKey` vector size (even though the consensus limit is much smaller at `~1MB` for outputs, `10KB` when spending as an input).
```

File: readme.md [hunk 3]
```diff
@@ -136,13 +136,22 @@ PREFIX_TOKEN <category_id> <has_nft> [commitment_length commitment] <ft_amount>
 
 </details>
 
-By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are advised to ensure that values between `253` (`0xfdfd00`) and `65535` (`0xfdffff`) can be parsed. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
+By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are REQUIRED to ensure that values between `253` (`0xfdfd00`) and `33554432` (`0xff00000002`) can be parsed, and that values greater than or equal to `33554433` (`0xff01000002`) are rejected with a parse failure. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
```
Comment (cculianu) [inline readme.md:139]:
```suggestion
By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are REQUIRED to ensure that values between `253` (`0xfdfd00`) and `33554432` (`0xff0000000200000000`) can be parsed, and that values greater than or equal to `33554433` (`0xff01000002`) are rejected with a parse failure. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
```

File: readme.md [hunk 4]
```diff
@@ -136,13 +136,22 @@ PREFIX_TOKEN <category_id> <has_nft> [commitment_length commitment] <ft_amount>
 
 </details>
 
-By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are advised to ensure that values between `253` (`0xfdfd00`) and `65535` (`0xfdffff`) can be parsed. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
+By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are REQUIRED to ensure that values between `253` (`0xfdfd00`) and `33554432` (`0xff0000000200000000`) can be parsed, and that values greater than or equal to `33554433` (`0xff01000002`) are rejected with a parse failure. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
```
Comment (cculianu) [inline readme.md:139]:
```suggestion
By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are REQUIRED to ensure that values between `253` (`0xfdfd00`) and `33554432` (`0xff0000000200000000`) can be parsed, and that values greater than or equal to `33554433` (`0xff0100000200000000`) are rejected with a parse failure. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
```

File: readme.md [hunk 5]
```diff
@@ -136,13 +136,22 @@ PREFIX_TOKEN <category_id> <has_nft> [commitment_length commitment] <ft_amount>
 
 </details>
 
-By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are advised to ensure that values between `253` (`0xfdfd00`) and `65535` (`0xfdffff`) can be parsed. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
+By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are REQUIRED to ensure that values between `253` (`0xfdfd00`) and `33554432` (`0xff0000000200000000`) can be parsed, and that values greater than or equal to `33554433` (`0xff0100000200000000`) are rejected with a parse failure. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
```
Comment (cculianu) [inline readme.md:139]:
```suggestion
By consensus, `commitment_length` is limited to `40` (`0x28`), but future upgrades may increase this limit. Implementers are REQUIRED to ensure that values between `253` (`0xfdfd00`) and `33554432` (`0xfe00000002`) can be parsed, and that values greater than or equal to `33554433` (`0xfe01000002`) are rejected with a parse failure. (See [Non-Fungible Token Commitment Length](#non-fungible-token-commitment-length).)
```

**Comment by bitjson** [comment]
Thanks for the PR! Makes sense – I need to review this a little more thoroughly, and I'll try to have it merged this week. 👍

**Comment by cculianu** [comment]
Closing. Don't merge this since it's obviated by the decision to not go with #41.

**Pull request closed without merging**